Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp

```diff
@@ -294,27 +256,34 @@ Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const GraphicsLayer& l
 
 Ref<CoordinatedTileBuffer> SkiaPaintingEngine::postPaintingTask(const GraphicsLayer& layer, RenderingMode renderingMode, const IntRect& dirtyRect, bool contentsOpaque, float contentsScale)
 {
+    auto buffer = createBuffer(renderingMode, dirtyRect.size(), contentsOpaque);
+    buffer->beginPainting();
+
     WTFBeginSignpost(this, RecordTile);
-    auto displayList = recordDisplayList(renderingMode, layer, dirtyRect, contentsOpaque, contentsScale);
+    auto* recordingCanvas = m_pictureRecorder.beginRecording(dirtyRect.width(), dirtyRect.height());
+    GraphicsContextSkia recordingContext(*recordingCanvas, renderingMode, RenderingPurpose::LayerBacking, nullptr, GraphicsContextSkia::ContextType::RecordingContext);
+    paintIntoGraphicsContext(layer, recordingContext, dirtyRect, contentsOpaque, contentsScale);
+    auto picture = m_pictureRecorder.finishRecordingAsPicture();
     WTFEndSignpost(this, RecordTile);
 
-    auto buffer = createBuffer(renderingMode, dirtyRect.size(), contentsOpaque);
-    buffer->beginPainting();
+    auto replayCanvas = SkiaReplayCanvas::create(dirtyRect.size(), recordingContext.releaseImageFences());
```

Comment:
I think it would be better to create the replay canvas in the worker thread, since it's only needed for replay, and only when rendering mode is accelerated, for cpu rendering we don't want it

Reply:
Ack. Already changed it, only think that needs to move cross-threads is the new `ImageFenceMap` (which is just a `using ImageFenceMap = HashMap<const SkImage, std::unique_ptr<GLFence>>`.